File: solutions/21/sol.js
const LIMIT = 10_000;

const getD = (n) => {
  const squareRoot = Math.floor(Math.sqrt(n));
  const properDivisors = [1];
  for (let num = 2; num < squareRoot; num++) {
    if (n % num === 0) {
      properDivisors.push(num);
      properDivisors.push(n / num);
    }
  }
  if (squareRoot >= 2) {
    if (squareRoot * squareRoot === n) {
      properDivisors.push(squareRoot)
    } else {
      if (n % squareRoot === 0) {
        properDivisors.push(squareRoot);
        properDivisors.push(n / squareRoot);
      }
    }
  }
  return properDivisors;
}

const getSum = (arr) => {
  let sum = 0;
  for (const num of arr) sum += num;
  return sum;
}

const getAnswer = (n) => {
  let sum = 0;
  const dObj = {}
  dObj[1] = 0;
  for (let num = 2; num < n; num++) {
    dObj[num] = getSum(getD(num));
  }
  for (let num = 2; num < n; num++) {
    const d1 = dObj[num];
    if (d1 === num) continue;
    let d2;
    if (d1 < n) {
      d2 = dObj[d1];
    } else {
      d2 = getSum(getD(d1));
    }
    if (d2 === num) sum += num;
  }
  return sum;
}

const answer = getAnswer(LIMIT);
console.log(answer);

File: solutions/45/sol.py
# 1*n*n + 1*n - 2*T(n) = 0
# 3*n*n - 1*n - 2*P(n) = 0
# H(n) = n*(2*n-1)

import math

def is_integer_number(x):
  return math.floor(x) == x

def solveQuadratic(a, b, c):
  d = b*b - 4*a*c
  if d < 0:
    return []
  if d==0:
    x = -b / (2*a)
    if is_integer_number(x):
      return [x]
    return []
  delta = math.sqrt(d)
  x1 = (-b + delta) / (2 * a)
  x2 = (-b - delta) / (2 * a)
  arr = []
  if is_integer_number(x1) and x1 > 0:
    arr.append(math.floor(x1))
  if is_integer_number(x2) and x2 > 0:
    arr.append(math.floor(x2))
  return arr

print(solveQuadratic(1, 1, -2 * 40755))
print(solveQuadratic(3, -1, -2 * 40755))